File: server/crates/domain/src/book_copy/logic.rs
use super::entity::{BookCopy, BookCopyCreationPayload, BookCopyPrepared};
use super::enums::BookCopyStatus;
use super::errors::BookCopyError;

impl BookCopy {
    /// Returns `true` when this copy is physically circulatable (not lost, not in maintenance).
    /// Full borrowability also requires checking that no active loan exists for this copy.
    #[must_use]
    fn is_circulatable(&self) -> bool {
        self.status == BookCopyStatus::Active
    }

    /// Guard: ensures copy is in a circulatable state for borrowing.
    pub fn ensure_circulatable(&self) -> Result<(), BookCopyError> {
        if !self.is_circulatable() {
            return Err(BookCopyError::CannotBeBorrowed);
        }
        Ok(())
    }

    /// Transition: Active -> Maintenance.
    pub fn send_to_maintenance(&self) -> Result<BookCopyStatus, BookCopyError> {
        if self.status != BookCopyStatus::Active {
            return Err(BookCopyError::CannotBeSentToMaintenance);
        }
        Ok(BookCopyStatus::Maintenance)
    }

    /// Transition: Maintenance -> Active.
    pub fn complete_maintenance(&self) -> Result<BookCopyStatus, BookCopyError> {
        if self.status != BookCopyStatus::Maintenance {
            return Err(BookCopyError::CannotBeReturnedFromMaintenance);
        }
        Ok(BookCopyStatus::Active)
    }

    /// Transition: any non-Lost -> Lost.
    pub fn mark_lost(&self) -> Result<BookCopyStatus, BookCopyError> {
        if self.status == BookCopyStatus::Lost {
            return Err(BookCopyError::CannotMarkBookLost);
        }
        Ok(BookCopyStatus::Lost)
    }

    /// Transition: Lost -> Active.
    pub fn mark_found(&self) -> Result<BookCopyStatus, BookCopyError> {
        if self.status != BookCopyStatus::Lost {
            return Err(BookCopyError::CannotBeReturnedFromLost);
        }
        Ok(BookCopyStatus::Active)
    }
}

impl BookCopyCreationPayload {
    #[must_use]
    pub fn prepare(self) -> BookCopyPrepared {
        BookCopyPrepared {
            barcode: self.barcode,
            book_id: self.book_id,
            status: BookCopyStatus::Active,
        }
    }
}

File: servers/go/internal/domain/bookcopy/logic.go
package bookcopy

func (c *BookCopy) EnsureCanBeBorrowed() error {
	if c.Status != BookCopyStatusActive {
		return ErrCannotBeBorrowed
	}
	return nil
}

func (c *BookCopy) SendToMaintenance() (BookCopyStatus, error) {
	if c.Status != BookCopyStatusActive {
		return "", ErrCannotBeSentToMaintenance
	}
	return BookCopyStatusMaintenance, nil
}

func (c *BookCopy) CompleteMaintenance() (BookCopyStatus, error) {
	if c.Status != BookCopyStatusMaintenance {
		return "", ErrCannotBeReturnedFromMaintenance
	}
	return BookCopyStatusActive, nil
}

func (c *BookCopy) MarkLost() (BookCopyStatus, error) {
	if c.Status == BookCopyStatusLost {
		return "", ErrCannotMarkBookLost
	}
	return BookCopyStatusLost, nil
}

func (c *BookCopy) MarkFound() (BookCopyStatus, error) {
	if c.Status != BookCopyStatusLost {
		return "", ErrCannotBeReturnedFromLost
	}
	return BookCopyStatusActive, nil
}

func (p BookCopyCreationPayload) Prepare() BookCopyPrepared {
	return BookCopyPrepared{
		Barcode: p.Barcode,
		BookID:  p.BookID,
		Status:  BookCopyStatusActive,
	}
}
